torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py

```diff
@@ -0,0 +1,253 @@
+      apply_fake_quant_fn (Callable): function that transforms `original_tensor` to fake quantized values
+    """
+
+    @staticmethod
+    def __new__(
+        cls,
+        original_tensor: torch.Tensor,
+        apply_fake_quant_fn: Callable,
+        fake_quant_enabled: bool = True,
+        **kwargs,
+    ):
+        kwargs.setdefault("dtype", original_tensor.dtype)
+        kwargs.setdefault("device", original_tensor.device)
+        return torch.Tensor._make_wrapper_subclass(
+            cls,
+            original_tensor.shape,
+            **kwargs,
+        )
+
+    def __init__(
+        self,
+        original_tensor: torch.Tensor,
+        apply_fake_quant_fn: Callable,
+        fake_quant_enabled: bool = True,
+        **kwargs
+    ):
+        self.original_tensor = original_tensor
+        self.apply_fake_quant_fn = apply_fake_quant_fn
+        self.fake_quant_enabled = fake_quant_enabled
+
+    def __tensor_flatten__(self):
+        return ["original_tensor"], [self.apply_fake_quant_fn, self.fake_quant_enabled]
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride,
+    ):
+        original_tensor = tensor_data_dict["original_tensor"]
+        (apply_fake_quant_fn, fake_quant_enabled) = tensor_attributes
+        return cls(
+            original_tensor,
+            apply_fake_quant_fn,
+            fake_quant_enabled,
+        )
+
+    @classmethod
+    def from_float(
+        cls,
+        input_float: torch.Tensor,
+        mapping_type: MappingType,
+        block_size: Tuple[int, ...],
+        target_dtype: torch.dtype,
+        quant_min: Optional[int] = None,
+        quant_max: Optional[int]  = None,
+        eps: Optional[float] = None,
+        scale_dtype: Optional[torch.dtype] = None,
+        zero_point_dtype: Optional[torch.dtype] = None,
+        preserve_zero: bool = True,
+        zero_point_domain: ZeroPointDomain = ZeroPointDomain.INT,
+    ):
+        def apply_fake_quant_fn(t: torch.Tensor):
+            assert isinstance(t, cls)
+            qmin, qmax = _get_and_check_qmin_qmax(target_dtype, quant_min, quant_max)
+            scale, zero_point = choose_qparams_affine(
+                t.original_tensor,
+                mapping_type,
+                block_size,
+                target_dtype,
+                qmin,
+                qmax,
+                eps,
+                scale_dtype,
+                zero_point_dtype,
+                preserve_zero,
+                zero_point_domain,
+            )
+            fq = _GenericFakeQuantize.apply(
+                t,
+                block_size,
+                scale,
+                zero_point,
+                qmin,
+                qmax,
+                zero_point_domain,
+            )
+            return fq
+        return cls(
+            input_float,
+            apply_fake_quant_fn,
+            fake_quant_enabled=True,
+            requires_grad=True,
```

Comment:
should we remove this and just get this from `input_float.requires_grad`?

Reply:
I feel it's better to always set it to True here since we always want gradients for QAT

Reply:
even that I feel just reusing the requires_grad from `input_float` might be beneficial, otherwise we'd have to worry about the inconsistency here, similar for other things like dtype, shape, device etc.

Reply:
Ok, relaxed this to use the original tensor's `requires_grad`